aten/src/ATen/native/cpu/PowKernel.cpp

```diff
@@ -35,62 +35,165 @@ void pow_tensor_tensor_kernel(TensorIterator& iter) {
 }
 
 void pow_tensor_scalar_kernel(TensorIterator& iter, Scalar exp_scalar) {
-  if (isFloatingType(iter.dtype())) {
-    const auto exp = exp_scalar.to<double>();
+  if (isFloatingType(iter.dtype())) {    
```

Comment:
pow is such a broken function, and this change greatly extends its length, making it even more unmaintainable.

Is there any way to not duplicate the code 3 times?

Also, if we pursue this change we should change the dispatch to be on iter.common_dtype(), not iter.dtype(). iter.dtype() is the output's dtype. iter.common_dtype() is the dtype the computation should be performed in.

Reply:
Yes, these changes have made the code quite unreadable.
If I specialize `pow_tensor_scalar_kernel()`, the code would still be duplicated, but it'd become more readable.
If I instead add conditionals in the existing code, its length would only increase by a few lines, and readability won't be as adversely affected.

Regarding  `iter.common_dtype()`, this one seems to be a blunder! Thanks for catching it!

Reply:
Let's try adding the conditionals to the existing code, and then we can look at that option more closely together.

Reply:
Hello @mruberry, I recollected that when I had initially started working on this PR, I had been unable to make the changes more concise because conditional statements raise invalid type conversion errors even if a condition won't be true upon runtime, as `scalar_t` is present in the code, which corresponds to a type. An example of such an error is `map(std::sqrt())`not supporting the `Half` type for AVX2 optimization. So, at most, I can combine the implementation for `float`, `double` & `bfloat16`, but there would still be some redundant code for `float16`.

An alternative is to use `pow_tensor_scalar_kernel()` as a wrapper for invoking the _actual_ kernels corresponding to a `common_dtype()` or group of types. For example, `pow_tensor_scalar_kernel_complex()` can be created from the current complex implementation, which can be called for `complex` types. Perhaps different dispatch stubs can then be defined & registered for them as well.

Please advise as to which of the two options would be better, or if another alternative comes to mind.
Thank you!

Reply:
Hello @mruberry,

Without a redesign, I think this duplication can't be avoided if conditional statements are used to reuse common-code:

1. `BFloat16` shares all but one line of its implementation with `float` & `double`. However, avoiding code-duplication by dispatching for these three types together would lead to overflow exceptions in some cases.
eg. consider `pow(Long tensor, Scalar double exponent)`. In this case, `common_dtype` &` iter.dtype()` would be `float`. We can't avoid trying to cast the `double` exponent into `float`, which would lead to an overflow. As to why a `double` would be casted into a `float` in the first place, please refer to the next few lines.

When only `float` & double `types` are dispatched together, as in the existing code on the main repo, we use:
`auto exp = exp_scalar.to<double>();`
However, if we dispatch `BFloat16`, `float` & `double` together, we can't use the above line of code, as compilation would fail with an error `cannot convert ‘const double’ to ‘const at::vec256::{anonymous}::Vec256<c10::BFloat16>&`.
.If we instead use `auto exp = exp_scalar.to<scalar_t>();`, we'd encounter an overflow exception in some use-cases, as in the example above (which is from a test in `pow_test.cpp`).

2. `Half`/`Float16` shares all but two conditional cases of values of `exp` with `BFloat16` and with `float` & `double`, so there seems to be some potential for decreasing code-duplication, but compilation would fail if we'd try to dispatch for `float16` & `bfloat16` together even with conditional statements because of the presence of `scalar_t` in the code, which would cause compilation errors because of AVX2 optimizations of `sqrt()` & `rsqrt()` not having been implemented for Half, even if we have something like
```
        if ((exp == 0.5) && (iter.common_dtype() != ScalarType::Half)) {
        // compilation fails because the compiler also tries to compile with scalar_t as a decltype for Half & fails
          cpu_kernel_vec(iter, 
            [=](scalar_t base) -> scalar_t {
              return std::sqrt(base);
            },
            [](Vec base) -> Vec { return base.sqrt(); }
          );
        }
```
Please advise if we should look at some other options, such as using `pow_tensor_scalar_kernel()` as a wrapper to invoke some methods that we can create for a type/group of types. In the meantime, I'll add an `OpInfo` for `pow()` & will enable more tests. Thank you!

Reply:
Thank you for this thorough explanation, @imaginary-person. I'll take a look (and maybe call in a partner who can take a very close look!).

Reply:
Thanks again for your help, @mruberry! In the meantime, I'd add small changes such as skipping `grad_pow `& `grad_grad_pow `tests, so that the only CI failures pertaining to OpInfo-enabled testing would be related to JIT-related tests, for which I had submitted issue #51786.

Reply:
I was wondering if `pow()` should have two OpInfos as well.

Reply:
Typically we don't need two OpInfos for tensor x tensor vs. tensor x scalar. In this case I don't think we need them because the operation being performed (pow) is still the same.